### PR TITLE
Mock Machine IDE (MemoryFsKite, MemoryFsKlient)

### PR DESCRIPTION
**Pull request by usirin**
This PR introduces a new concept to create an IDE with a mock klient/kite instances. It uses the `memory-fs` npm module to create a fs like object in memory, and provides the same api as node.js `fs` module.

This is really a bare bones implementation to see if it's feasible for tests (especially for rainforest) to use, but we can use this as a base fs kite implementation, and build up on it.

/cc @devrim @diclee 

Review can start I am done with 98% of the implementation.

### CODE REVIEW

**Comment by usirin** [comment]
@gokmen ready to go.

Here is what it does:

- Hit the `/IDE/test-machine` or `/IDE/test-machine/test-workspace`
- Hit `RUN` button on modal.
- It runs the tests in `tests` array in `/client/ide/lib/test/index.coffee` [here](https://github.com/koding/koding/blob/mock-kite-klient/client/ide/lib/test/browser/index.coffee#L9-L11)
- An example test is located under `/client/ide/lib/test/browser.coffee`

## Screencast

mock-klient

File: client/app/lib/providers/machine.coffee [hunk 1]
```diff
@@ -114,6 +122,11 @@ module.exports = class Machine extends KDObject
 
     { kontrol } = kd.singletons
 
+    # this is a chance for other middlewares to inject their own kite/klient.
+    klient = runMiddlewares.sync this, 'getBaseKite', createIfNotExists
+
+    return klient  if klient
```
Comment (gokmen) [inline client/app/lib/providers/machine.coffee]: let's change this one to;

```coffee
testklient = runMiddlewares.sync this, 'getBaseKite', createIfNotExists
return testklient  if testklient
```

File: client/app/lib/providers/middlewares/testmachine.coffee [hunk 2]
```diff
@@ -0,0 +1,40 @@
+{ assign } = require 'lodash'
+isTestMachine = require 'app/util/isTestMachine'
+MemoryFsKlient = require 'app/kite/kites/memoryfsklient'
+
+
+module.exports = TestMachineMiddleware =
```
Comment (gokmen) [inline client/app/lib/providers/middlewares/testmachine.coffee]: 👍

Comment (gokmen) [inline client/app/lib/providers/middlewares/testmachine.coffee]: a brief comment would be nice

File: client/ide/lib/index.coffee [hunk 3]
```diff
@@ -622,31 +622,22 @@ class IDEAppController extends AppController
...
 
       unless machineItem
         return @showNoMachineState()
 
       unless machineItem instanceof Machine
         machineItem = new Machine { machine: machineItem }
 
-      # Don't run these lines on `Teams` scope.
-      # Because `Teams` uses new Sidebar with React + Flux
-      if not isTeamReactSide() and not machineItem.isMine() and not machineItem.isApproved()
```
Comment (gokmen) [inline client/ide/lib/index.coffee]: 👊

File: client/ide/lib/test/browser/example.coffee [hunk 4]
```diff
@@ -0,0 +1,10 @@
+$ = require 'jquery'
+
+module.exports = (expect) ->
+
+  # select finder item for `.bash_rc`
+  fileItemSelector = 'span[title*=bash_rc]'
```
Comment (gokmen) [inline client/ide/lib/test/browser/example.coffee]: 😕

File: client/ide/lib/test/browser/index.coffee [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    message: message
+    buttonTitle: 'Close'
+    onButtonClick: -> modal.destroy()
+
+# /cc @gokmen: this might be a strong possible replacement for
+# pistachio. KDViews will be where we write logic, React components will always
+# be small enough to be treated as a pistachio string.
+# This way, each ReactView wrapping kd view can implement what the f*ck it
+# wants, either bongo model, or a redux store. KDViews could be the solutions
+# of colocating data with view components.
```
Comment (gokmen) [inline client/ide/lib/test/browser/index.coffee]: 👍

File: client/app/lib/kite/kites/memoryfsklient.coffee [hunk 6]
```diff
@@ -0,0 +1,76 @@
+kd = require 'kd'
+Promise = require 'bluebird'
+KodingKite = require '../kodingkite'
+Klient = require './klient'
+nick = require 'app/util/nick'
+MemoryFsKite = require './memoryfs'
+
+module.exports = class MemoryFsKlient extends KodingKite
```
Comment (gokmen) [inline client/app/lib/kite/kites/memoryfsklient.coffee]: 👍

**Review by gokmen — CHANGES_REQUESTED** [review]
Looks great in general just some small changes needed, awesome work 💯

File: client/app/lib/kite/kodingkite.coffee [hunk 7]
```diff
@@ -29,6 +31,8 @@ module.exports = class KodingKite extends kd.Object
 
     { name } = options
 
+    @commandBuffer = new KiteCommandBuffer { kite: this }
```
Comment (usirin) [inline client/app/lib/kite/kodingkite.coffee]: 👍 You're right.

File: client/app/lib/kite/kodingkite.coffee [hunk 8]
```diff
@@ -91,13 +91,12 @@ module.exports = class KodingKite extends kd.Object
 
             resolve (@transport?.tell args...
 
-              .then (res) ->
+              .then (res) =>
```
Comment (gokmen) [inline client/app/lib/kite/kodingkite.coffee]: no need for fat arrow here.

Comment (gokmen) [inline client/app/lib/kite/kodingkite.coffee]: better to revert any change on this file actually.

File: client/app/lib/userenvironmentdataprovider.coffee [hunk 9]
```diff
@@ -87,6 +101,13 @@ module.exports =
     @getAllMachines().filter (vm) -> vm.machine.status.state is 'Running'
 
 
+  getMachineWithPredicate: (fn, predicate) ->
+
+    filtered = @getAllMachines().filter predicate
+
+    if filtered.length then filtered[0] else null
```
Comment (gokmen) [inline client/app/lib/userenvironmentdataprovider.coffee]: `@getAllMachines().filter(predicate)[0]` ?

**Review by gokmen — CHANGES_REQUESTED** [review]
just small changes, rest LGTM 💯

File: client/app/lib/userenvironmentdataprovider.coffee (hunk 9 above)
Comment (usirin) [inline client/app/lib/userenvironmentdataprovider.coffee]: 👍

**Comment by devrim** [comment]
awesome job @usirin - @hakankaradis can you create a new issue that outlines the keyb shortcut tests (and maybe others, consult to @diclee ad @szkl) and get it done by an external developer?

**Review by gokmen — APPROVED** [review]

**Pull request merged**